wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/RdsMultiAzDbClusterListProvider.java

```diff
@@ -0,0 +1,197 @@
+  private final String fetchWriterNodeQuery;
+  private final String fetchWriterNodeQueryHeader;
+  static final Logger LOGGER = Logger.getLogger(RdsMultiAzDbClusterListProvider.class.getName());
+
+  public RdsMultiAzDbClusterListProvider(
+      final Properties properties,
+      final String originalUrl,
+      final HostListProviderService hostListProviderService,
+      final String topologyQuery,
+      final String nodeIdQuery,
+      final String isReaderQuery,
+      final String fetchWriterNodeQuery,
+      final String fetchWriterNodeQueryHeader
+  ) {
+    super(properties,
+        originalUrl,
+        hostListProviderService,
+        topologyQuery,
+        nodeIdQuery,
+        isReaderQuery);
+    this.fetchWriterNodeQuery = fetchWriterNodeQuery;
+    this.fetchWriterNodeQueryHeader = fetchWriterNodeQueryHeader;
+  }
+
+  /**
+   * Obtain a cluster topology from database.
+   *
+   * @param conn A connection to database to fetch the latest topology.
+   * @return a list of {@link HostSpec} objects representing the topology
+   * @throws SQLException if errors occurred while retrieving the topology.
+   */
+  protected List<HostSpec> queryForTopology(final Connection conn) throws SQLException {
+
+    int networkTimeout = -1;
+    try {
+      networkTimeout = conn.getNetworkTimeout();
+      // The topology query is not monitored by the EFM plugin, so it needs a socket timeout
+      if (networkTimeout == 0) {
+        conn.setNetworkTimeout(networkTimeoutExecutor, defaultTopologyQueryTimeoutMs);
+      }
+    } catch (SQLException e) {
+      LOGGER.warning(() -> Messages.get("RdsHostListProvider.errorGettingNetworkTimeout",
+          new Object[] {e.getMessage()}));
+    }
+
+    try {
+      final Statement stmt = conn.createStatement();
+      String writerNodeId = processWriterNodeId(stmt.executeQuery(this.fetchWriterNodeQuery));
+      if (writerNodeId.isEmpty()) {
+        final ResultSet nodeIdResultSet = stmt.executeQuery(this.nodeIdQuery);
+        while (nodeIdResultSet.next()) {
+          writerNodeId = nodeIdResultSet.getString(1);
+        }
+      }
+      final ResultSet topologyResultSet = stmt.executeQuery(this.topologyQuery);
+      return processTopologyQueryResults(topologyResultSet, writerNodeId);
+    } catch (final SQLSyntaxErrorException e) {
+      throw new SQLException(Messages.get("RdsHostListProvider.invalidQuery"), e);
+    } finally {
+      if (networkTimeout == 0 && !conn.isClosed()) {
+        conn.setNetworkTimeout(networkTimeoutExecutor, networkTimeout);
+      }
+    }
+  }
+
+  /**
+   * Get writer node ID.
+   *
+   * @param fetchWriterNodeResultSet A ResultSet of writer node query
+   * @return String The ID of a writer node
+   * @throws SQLException if errors occurred while retrieving the topology
+   */
+  private String processWriterNodeId(final ResultSet fetchWriterNodeResultSet) throws SQLException {
+    String writerNodeId = "";
+    while (fetchWriterNodeResultSet.next()) {
```

Comment:
is it possible that there are two writers?

Reply:
Not something I'm aware of, see https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/multi-az-db-clusters-concepts.html

Reply:
if that is the case why do we have a loop here ?